File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.java
import java.io.*;
import java.util.Arrays;

public class Main{
    public static void main(String[] args) throws IOException{

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());
        int[][] tri = new int[n][];
        int[][] dp = new int[n][];

        for(int i=0; i<n; i++) {
            int[] arr = new int[i+1];
            dp[i] = arr;
        }
        for (int i=0; i<n; i++) {
            String[] input = br.readLine().split(" ");
            int[] arr = new int[i+1];
            for(int j=0; j<input.length; j++)
                arr[j] = Integer.parseInt(input[j]);

            tri[i] = arr;
        }

        dp[0][0] = tri[0][0];

        for(int i=1; i<tri.length; i++) {
            for(int j=0; j<tri[i].length; j++) {
                if(j == 0){
                    dp[i][j] = dp[i-1][j] + tri[i][j];
                }
                else if(j == tri[i].length - 1) {
                    dp[i][j] = dp[i-1][j-1] + tri[i][j];
                }
                else {
                    dp[i][j] = Math.max(dp[i-1][j-1], dp[i-1][j]) + tri[i][j];
                }
            }
        }

        bw.write(Arrays.stream(dp[n-1]).max().getAsInt() +"");
        bw.flush();
    }
}

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.py
if __name__ == '__main__':

    INF = int(1e9)
    n = int(input())

    tri = []

    for _ in range(n):
        tri.append(list(map(int, input().split())))

    dp = [[INF] * (i+1) for i in range(n)]
    dp[0][0] = tri[0][0]

    for i in range(1, n):
        for j in range(len(tri[i])):
            if j == 0:
                dp[i][j] = tri[i][j] + dp[i - 1][j]
            elif j == len(tri[i]) - 1:
                dp[i][j] = tri[i][j] + dp[i - 1][j - 1]
            else:
                dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j]) + tri[i][j]
    print(max(dp[n - 1]))